### ISSUE TITLE
Self Hosted ASP.NET Core 2.0: TagHelper doesnt work

### ISSUE THREAD

**Issue by roeb** [aspnet/Home#2216]
I have ASP.NET Core 2.0 website, which is hosted in a .NET Core console / service. I have implemented a easy custom TagHelper. The TagHelper works fine, when I started the ASP.NET Core website directly from VS2017. When i start the .NET Core console host, which has hosted the asp.net page, the tag helper doesnt work.

I have try to register the TagHelper in the _ViewImports.cshtml and also directly in the view as self (Home/Index).

My custom TagHelper:
```C#
[HtmlTargetElement("time")]
public class TimeTagHelper : TagHelper
{
    public override async Task ProcessAsync(TagHelperContext context, TagHelperOutput output)
    {
        output.Content.SetHtmlContent($"<h1>{DateTime.Now.ToShortTimeString()}</h1>");
    }
}
```
_ViewImports.cshtml
```C#
@using SelfHosted.Website
@addTagHelper *, Microsoft.AspNetCore.Mvc.TagHelpers
@addTagHelper SelfHosted.Website.TagHelpers.*, SelfHosted.Website
```
The TimeTagHelper intellisense in VS2017 work fine: [TimeTagHelper in Index.cshtml](https://i.stack.imgur.com/LoyyQ.png)

The code is on [GitHub](https://github.com/roeb/DotNetCoreSamples). The TagHelper is implemented in SelfHosted.Website and used in /Home/Index. It works perfect when the SelfHosted.Website started directly. But it doesnt work, when i start the Website from SelfHostes.Console.

Any ideas? :)

Thanks Robert

**Comment by Tratcher** [comment]
SelfHosted.Console is a WindowsService? We have a writeup for that. https://docs.microsoft.com/en-us/aspnet/core/hosting/windows-service. I wonder if the content root path is wrong so parts of your app aren't being discovered.

**Comment by roeb** [comment]
Correct. SelfHosted.Console is a WindowsService but it can also started as standalone from the commandline. It works pretty good, except for the TagHelper.
All views and all other content in wwwroot is embedded content in the SelfHosted.Website Assembly. The ContentRoot is set to the Root folder from the SelfHosted.Console (where the dll from the Selfhosted.Website is).

```C#
    public class WebModule
    {
        private IWebHost _host;

        public Task StartAsync()
        {
            _host = new WebHostBuilder()
               .UseKestrel()
               .UseUrls("http://localhost:5003")
               .UseContentRoot(Path.GetDirectoryName(GetType().Assembly.Location))
               .UseIISIntegration()
               .UseStartup<Startup>()
               .Build();

            _host.Run();

            return Task.CompletedTask;
        }
    }
``` 
It surprised me that everything worked (css, scripts and views), except for the TagHelper.

**Comment by Tratcher** [comment]
@NTaylorMullen

**Comment by NTaylorMullen** [comment]
@roeb in order to make your scenario work you need to accomplish two things:
1. Since you're having a different host boot your application you need to copy your applications deps.json into your console apps directory. This is how MVC determines which compilation assemblies to use at runtime. 
2. You're not embedding the `_ViewImports.cshtml` file. Your console app is depending on the application it boots having its views embedded; therefore,  TagHelpers wont be `@addTagHelper`d until the `_ViewImports` can be found.

Fixing these two things makes everything work 👍

**Comment by roeb** [comment]
Thanks @NTaylorMullen that works for me :ok_hand:

These are the steps that helped me:

1.  Add the following ItemGroup definition to my Selhosted.Website.csproj
```xml
 <ItemGroup>
    <EmbeddedResource Include="wwwroot\**\*;Views\**\*;Areas\**\Views" />
  </ItemGroup>
```
2. Add the ```PreserveCompilationContext``` property to the PropertyGroup in  Selhosted.Website.csproj
```xml
 <PropertyGroup>
    <TargetFramework>netcoreapp2.0</TargetFramework>
    <AssemblyName>SelfHosted.Website</AssemblyName>
    <RootNamespace>SelfHosted.Website</RootNamespace>
    <TypeScriptToolsVersion>2.3</TypeScriptToolsVersion>
    <PreserveCompilationContext>true</PreserveCompilationContext>
  </PropertyGroup>
```
3. Include the generated SelfHosted.Website.deps.json from the websites bin folder to the SelfHosted.Console project and set the Copy to Output Directory to Copy always.

![SolutionPicture](https://tc1uqw-db3pap001.files.1drv.com/y4mUgpMknXmLagCSjFe951m6tEmxo92KFyW6Gnu-saCeB6ddB-ra_WsCqz2QkyHEfTE7ID6EK_J-bo9Zu5p2B8X9IPVVUJLWUhoy5rNmSihytCQDsiL9vxcTqtx-K7bDEyz1b0Tc52A3eiCJNddAx5_ZPEbz8KIpn7dh4obSJ-d6QRPAIGS6y0NtLV51zuiY0BbuY4El7USjUQuWFifwawucQ?width=445&height=477&cropmode=none)

**Comment by roeb** [comment]
Hi @NTaylorMullen,

thanks for your help some time ago. Its works pretty fine for me, if I have referenced the project with the website directly as Project-Reference to my host. 

But now I'm in the situation, that I must reference the Project with the website as a nuget package to the host project. 
I have added the deps.json from the web project in the Host and also in the web project as embedded ressource and also with copy to output. But the self hosted website can't be find the ASP.NET Core Taghelpers. It works fine, when I referencing the project directly as project reference. 

Have you any idea how I could solve it?
